### PR TITLE
Redo termination logic for network threads, and cleanup NetworkManager

### PR DESCRIPTION
**Pull request by tli2**
In this PR I am putting the initialization logic for NetworkConnection (now named ConnectionHandle) into a factory class and changed the NetworkManager class accordingly (now called PelotonServer).

I am also able to get rid of the existing start/stop logic of the server as a result of those refactor.

I am not able to find a way to quickly break up ConnectionHandle into smaller, more cohesive classes though. So I will probably leave TO-DOs, and send in one last PR for error handling later today and call this refactor done.

### CODE REVIEW

File: src/include/network/connection_handle_factory.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    auto it = reusable_handles_.find(conn_fd);
+    if (it == reusable_handles_.end()) {
+      // We are not using std::make_shared here because we want to keep ConnectionHandle constructor
+      // private to avoid unintentional use.
+      auto handle = std::shared_ptr<ConnectionHandle>(
+          new ConnectionHandle(conn_fd, handler, std::make_shared<Buffer>(), std::make_shared<Buffer>()));
+      reusable_handles_[conn_fd] = handle;
+      return handle;
+    }
+
```
Comment (dengyuchenkit) [inline src/include/network/connection_handle_factory.h]: I know it's not part of this PR but am having some confusions here. Are all the ConnectionHandles exist in the `reusable_handles_` pool? And Is it also keeping those ConnectionHandles whose connections are already closed but these objects are not destroyed yet so we can reuse them later?

File: src/include/network/connection_handle_factory.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "network/connection_handle.h"
+#include "peloton_server.h"
+
+namespace peloton {
+  namespace network {
+
```
Comment (dengyuchenkit) [inline src/include/network/connection_handle_factory.h]: Extra whitespace.

File: src/include/network/connection_handle.h [hunk 3]
```diff
@@ -12,18 +12,81 @@
...
+ * @see ConnectionHandleFactory
+ */
+class ConnectionHandle {
 public:
+
+  /**
+   * refill_read_buffer - Used to repopulate read buffer with a fresh
+   * batch of data from the socket
+   */
+  Transition FillReadBuffer();
```
[the remark this answers is not in the document]
Comment (tli2) [reply src/include/network/connection_handle.h]: Yes. Will do.

File: src/include/network/connection_handle_factory.h (hunk 1 above)
Comment (tli2) [reply src/include/network/connection_handle_factory.h]: Yes.

File: src/include/network/peloton_server.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  // static void LogCallback(int severity, const char *msg);
+
+  uint64_t port_;             // port number
+  int listen_fd_ = -1;         // server socket fd that PelotonServer is listening on
+  size_t max_connections_;    // maximum number of connections
+
+  std::string private_key_file_;
+  std::string certificate_file_;
+
+  std::shared_ptr<ConnectionDispatcherTask> dispatcher_task;
```
[the remark this answers is not in the document]
Comment (tli2) [reply src/include/network/peloton_server.h]: Yes.

**Pull request merged**